src/modalities/nn/moe.py

```diff
@@ -0,0 +1,141 @@
+from typing import Callable, Optional, Tuple
+
+import torch
+import torch.nn as nn
+from pydantic import BaseModel
+
+
+# MoE implementation inspired from Dbrx https://github.com/databricks/dbrx/blob/main/model/modeling_dbrx.py
+class MoEFFNConfig(BaseModel):
+    moe_num_experts: int
+    moe_top_k: int
+    moe_normalize_expert_weights: float
+    uniform_expert_assignment: bool
+    ffn_hidden_size: int
+    act_fn: Callable[[], nn.Module] = nn.SiLU
+    moe_jitter_eps: Optional[float]
+
+
+class MoERouter(nn.Module):
+    def __init__(self, hidden_size: int, moe_config: MoEFFNConfig):
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.moe_num_experts = moe_config.moe_num_experts
+        self.moe_top_k = moe_config.moe_top_k
+        self.moe_normalize_expert_weights = moe_config.moe_normalize_expert_weights
+        self.uniform_expert_assignment = moe_config.uniform_expert_assignment
+        self.moe_jitter_eps = moe_config.moe_jitter_eps
+
+        self.layer = nn.Linear(self.hidden_size, self.moe_num_experts, bias=False)
+
+    def forward(self, x: torch.Tensor) -> Tuple[torch.Tensor, torch.LongTensor]:
+        if self.training and self.moe_jitter_eps is not None:
+            x = x * self._jitter(x)
+
+        weights = self.layer(x.view(-1, x.shape[-1])).softmax(dim=-1, dtype=torch.float32)
+        top_weights, top_experts = torch.topk(weights, self.moe_top_k, dim=-1)
+
+        if self.moe_normalize_expert_weights:
+            top_weights = top_weights / torch.norm(
+                top_weights, p=self.moe_normalize_expert_weights, dim=-1, keepdim=True
+            )
+
+        if self.uniform_expert_assignment:
+            with torch.no_grad():
+                uniform_tensor = (
+                    torch.arange(0, top_experts.numel(), device=top_experts.device, dtype=top_experts.dtype)
+                    % self.moe_num_experts
+                )
+                top_experts = uniform_tensor.reshape(top_experts.shape)
+                # Note, weights and top_weights are not changed
+
+        top_weights = top_weights.to(x.dtype)
+        return top_weights, top_experts
+
+    def _jitter(self, x: torch.Tensor) -> torch.Tensor:
+        if self.moe_jitter_eps is None:
+            raise RuntimeError("The router does not have moe_jitter_eps set.")
+        low = 1.0 - self.moe_jitter_eps
+        high = 1.0 + self.moe_jitter_eps
+        noise = torch.rand(x.size(), dtype=x.dtype, device=x.device)
+        return low + noise * (high - low)
+
+
+class MoEExpertGLU(nn.Module):
```

Comment:
GLU = Gated Linear Unit ?
Does that mean each Expert is a GLU layer?
Why is it not just e.g. a FF Layer?

Reply:
Probably the naming choice was not the best on our part, GLU is for [Gated Linear Unite] (https://paperswithcode.com/method/glu).  the operation x1 = x1 * x2 is the gating mechanism of the GLU, with x2 being the gate, which control how much information/features form x1 are retained. I am speculating here, but i think this makes the training more stable , and nicely ties into the concept that each expert "specializes" on different features.